### DISCUSSION TITLE
關於 lab4-2 Firmware

### DISCUSSION THREAD

**Discussion by Josephood7** [bol-edu/HLS-SOC-Discussions#127] [Q&A]
關於Firmware
1. 請問Risc-v的Firmware控制X, Y in/out要寫在哪裡呢?或是已經寫好了的話是寫在哪裡
2. 4-2是只要寫兩個usr_proj_example.v就好嗎?要寫C的Firmware嗎?那個fir.c跟fir.h就是4-1寫好的嗎?還是要另外寫呢?
3. 4-2是兩種不同的FIR一起運作嗎?還是Firmware的FIR那邊是寫控制X, Y in/out的地方呢?

**Comment by LingChiYang** [comment]
1. 可參考include的file
https://github.com/bol-edu/caravel-soc_fpga-lab/blob/main/lab-caravel_fir/firmware/caravel.h
其實就是寫到0x30000000後的位址(X, Y)並compile成loadword或storeword

2. 4-2需要把4-1寫的RTL用user_proj_wrapper包起來。fir.c和fir.h只在4-1用，在4-2會在執行一次4-1的東西，因此跑完4-1的東西後剩餘的C code需要自己定義。

3. 先執行完4-1的FIR在執行4-2的FIR

以上同學再確認一下我有沒有回答到你的問題

**Reply by Josephood7** [reply]
謝謝您詳盡的回覆
關於第二點 不太懂剩餘的C code這邊是只要定義什麼 然後在哪裡定義

**Reply by LingChiYang** [reply]
在counter_la_fir.c定義即可，如果在別的地方定義，再請同學在report中註明

**Reply by v00000v** [reply]
請問user_proj_wrapper是包4-1的RTL以及Lab3的fir.v與bram嗎(如老師上課的說明)
![image](https://github.com/bol-edu/HLS-SOC-Discussions/assets/123727462/5ea4ab34-33a1-4265-bcff-a6d23cced52f)
